test/fail_compilation/test22574.d

```diff
@@ -0,0 +1,13 @@
+//https://issues.dlang.org/show_bug.cgi?id=22574
+/*
+TEST_OUTPUT:
+---
+fail_compilation/test22574.d(100): Error: variable `x` is used as a type
+fail_compilation/test22574.d(100):        variable `x` is declared here
```

Comment:
Can we agree that this is a weird error message? When looking at `x* x` the compiler should first try to analyze the type and then the variable name, however, in this case it seems that the name is inserted in the symbol table before actually resolving the type. I would expect that the error message is something along the lines of "type `x` is not defined".

Reply:
Looking in [template declaration semantic](https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L2560) you can see that when the parameter is [declared](https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L5534) (for template value parameters) the identifier is inserted in the scope before actually doing any semantic on the type. However, the next thing that it is done in [template declaration semantic](https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L2566) is to do [semantic on the type](https://github.com/dlang/dmd/blob/master/src/dmd/templateparamsem.d#L78). So, if we want to fix the underlying issue and report a correct error, we should add

```d
if (valtype)
    valtype = valtype.typeSemantic(loc, sc);
```

in TemplateValueParameter::declareParameter.

That will make the segfault go away and output: "Undefined identifier `x`"

Reply:
This is why I left the comment in the source code.